### PR TITLE
Optimization of SGs

### PR DESCRIPTION
**Pull request by YairSlobodin1**
- [x] read SGs (input)
- [x] write output
- [x] set all flags correctly
- [x] basic algorithm
- [x] tests
- [x] update `Readme.md`

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: cmd/subcmds/optimizeSG.go [hunk 1]
```diff
@@ -5,18 +5,40 @@ SPDX-License-Identifier: Apache-2.0
...
 	return cmd
 }
+
+func optimization(cmd *cobra.Command, args *inArgs) error {
+	cmd.SilenceUsage = true // if we got this far, flags are syntactically correct, so no need to print usage
+	sgs, err := confio.ReadSGs(args.configFile)
+	if err != nil {
+		return fmt.Errorf("could not parse config file %v: %w", args.configFile, err)
+	}
+	if optimize.ReduceSGRules(sgs, args.sgName) != nil {
```
[the remark this answers is not in the document]
Comment (YairSlobodin1) [reply cmd/subcmds/optimizeSG.go:40]: fixed in [#193](https://github.com/np-guard/vpc-network-config-synthesis/blob/output/cmd/subcmds/optimize.go#L32-L42)

File: cmd/subcmds/root.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ const (
 	outputFileFlag = "output-file"
 	outputDirFlag  = "output-dir"
 	prefixFlag     = "prefix"
+	sgNameFlag     = "sg-name"
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply cmd/subcmds/root.go:22]: I would say yes.

[the remark this answers is not in the document]
Comment (YairSlobodin1) [reply cmd/subcmds/root.go:22]: fixed in #193

File: README.md [hunk 3]
```diff
@@ -1,45 +1,70 @@
...
 * `vpcgen synth acl --single` - generate a single nACL for all subnets in the same VPC.
+* `vpcgen optimize sg` - optimize SGs.
+* `vpcgen optimize acl` - optimize nACLs (In progress).
 
-### nACLs Generation
-Specifying the `--single` flag results in generating a single nACL for all subnets in the same VPC. Otherwise, an nACL is generated for each subnet separately.  
-**Note**: A required connection between NIFs/VSIs/VPEs implies connectivity will be allowed between the subnets they are contained in.
+## Synthesis
+#### nACLs Generation 
+a required connection between NIFs/VSIs/VPEs implies connectivity will be allowed between the subnets they are contained in.
```
Comment (zivnevo) [inline README.md:17]:
```suggestion
A required connection between NIFs/VSIs/VPEs implies connectivity will be allowed between the subnets they are contained in.
```

File: README.md [hunk 4]
```diff
@@ -1,45 +1,70 @@
...
 
-### Output
-1. If the `output-dir` flag is used, the specified folder will be created, containing one file per VPC. Each generated file will contain the network resources (Security Groups or Network ACLs) relevant to its VPC. File names are set as `prefix_vpc`, where prefix is ​​the value received in the `prefix` flag. If the `prefix` flag is omitted, file names will match VPC names.
-2. If the `output-file` flag is used, all generated resources will be written to the specified file.
-3. if both `output-file` and `output-dir` flags are not used, the collection will be written to stdout.
+#### Options
+```commandline
+Flags:
+  -h, --help                help for synth
+  -s, --spec string         JSON file containing spec file
```
Comment (zivnevo) [inline README.md:30]:
```suggestion
Flags:
  -s, --spec string         JSON file containing spec file
```

File: README.md
Comment (zivnevo) [inline README.md:1]: Add a run example for `vpcgen optimize sg` in the last section

**Review by zivnevo — CHANGES_REQUESTED** [review]

**Review by zivnevo — APPROVED** [review]

**Pull request merged**